ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs

```diff
@@ -408,59 +408,51 @@ data InvalidBlockInfo blk = InvalidBlockInfo
 type FutureBlocks blk = Map (HeaderHash blk) (Header blk)
 
 {-------------------------------------------------------------------------------
-  Blocks to add
+  Chain selection queue
 -------------------------------------------------------------------------------}
 
--- | FIFO queue used to add blocks asynchronously to the ChainDB. Blocks are
--- read from this queue by a background thread, which processes the blocks
--- synchronously.
-newtype BlocksToAdd m blk = BlocksToAdd (TBQueue m (BlockToAdd m blk))
-  deriving NoUnexpectedThunks via OnlyCheckIsWHNF "BlocksToAdd" (BlocksToAdd m blk)
-
--- | Entry in the 'BlocksToAdd' queue: a block together with the 'TMVar's used
--- to implement 'AddBlockPromise'.
-data BlockToAdd m blk = BlockToAdd
-  { blockToAdd            :: !blk
-  , varBlockWrittenToDisk :: !(StrictTMVar m Bool)
-    -- ^ Used for the 'blockWrittenToDisk' field of 'AddBlockPromise'.
-  , varBlockProcessed     :: !(StrictTMVar m (Point blk))
+-- | FIFO queue used to perform chain selection on blocks asynchronously.
+-- Blocks are read from this queue by a background thread, which performs
+-- chain selection for each block synchronously.
+newtype ChainSelectionQueue m blk = ChainSelectionQueue (TBQueue m (BlockToProcess m blk))
```

Comment:
To automatically reorder blocks that arrived out of order, we could, for example, turn this into a `(T)(M)Var` containing an `OrdPSQ` ordered on the `SlotNo`.

Reply:
Actually, something simple like a `Map` (ordered by the `RealPoint`s) or just a list + `insertBy` is already enough.